### ISSUE TITLE
False Positive: <li> inside <menu role="toolbar"> triggers "List item parent element has a role that is not role='list'"

### ISSUE THREAD

**Issue by micaavigliano** [dequelabs/axe-core#5073]
### Product

axe Extension

### Product Version

4.126.2

### Latest Version

- [x] I have tested the issue with the latest version of the product

#### Expectation

The `<menu>` elements is an HTML element defined as a semantic alternative to `<ul>` for interactive items. According to the [W3C HTML Living Standard](https://html.spec.whatwg.org/multipage/grouping-content.html#the-menu-element) and [MDN](https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/menu#specifications), a `<menu>` element validly accepts `<li>` elements as direct children.

I expected axe-core to recognize the `<menu>` element as a valid parent for `<li>` tags, even when the `<menu>` is assigned a functional ARIA role like `role="toolbar"`, without triggering structural integrity errors.

#### Actual
Axe-core flags a Serious impact error: `<li>` elements must be contained in a `<ul>` or `<ol>`.
The tool specifically fails the audit because the parent element (`<menu>`) has a role other than `role="list"`. This results in multiple false positives (22 errors in my specific case) for markup that is semantically and syntactically correct according to HTML specifications.

#### How to Reproduce
Run an axe-core audit on the following HTML structure:

```html
<menu role="toolbar" aria-label="Text formatting">
  <li style="margin:0">
    <button type="button" aria-label="Bold" aria-pressed="false">B</button>
  </li>
  <li style="margin:0">
    <button type="button" aria-label="Italic" aria-pressed="false">I</button>
  </li>
  <li style="margin:0">
    <button type="button" aria-label="Underline" aria-pressed="false">U</button>
  </li>
</menu>
```

#### Additional context
This issue forces developers to either use non-semantic `<div>` tags or ignore serious accessibility flags, which is counter-intuitive for accessibility-first development.

**Comment by straker** [comment]
Thanks for the issue. I also see the code you provided flag it as a violation when it shouldn't. Looking at the code it looks like [listitem-evaluate](https://github.com/dequelabs/axe-core/blob/develop/lib/checks/lists/listitem-evaluate.js) only looks at the node name and explicit role of the element, but doesn't look at the resolved role so `menu` which is [role of `list`](https://github.com/dequelabs/axe-core/blob/develop/lib/commons/standards/implicit-html-roles.js#L189) isn't considered.

**Comment by straker** [comment]
Wait. I mispoke. the last line of the evaluate function looks at menu for the node name. What's really going on is that the menu has an explicit role of `toolbar` and so fails. This is correct behavior as a toolbar is not a valid parent of a listitem. A [toolbar](https://www.w3.org/TR/wai-aria-1.2/#toolbar) is similar to a menubar role which is different than the `menu` HTML element [which has the implicit role of `list`](https://www.w3.org/TR/html-aria/#el-menu)

**Comment by micaavigliano** [comment]
@straker  so how do I create a semantic toolbar if I want to be semantically correct? I know that `<menu>` has a semantic value of a `list` for the accessibility tree. The HTML element semantically accepts the `role="toolbar"`. https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/menu#technical_summary and https://html.spec.whatwg.org/multipage/grouping-content.html#the-menu-element I do not believe I am just inventing things.
In my perspective, this is an issue because "No ARIA is better than Bad ARIA". 

Literally the `<menu>` definition on whatwg

> The [menu](https://html.spec.whatwg.org/multipage/grouping-content.html#the-menu-element) element is simply a semantic alternative to [ul](https://html.spec.whatwg.org/multipage/grouping-content.html#the-ul-element) to express an unordered list of commands (a "toolbar"). 

On the accessibility tree, the children of the `<menu role="toolbar">` are automatically being detected as the interactive elements, meaning that the `<li>`s are being avoided because a new role override the native role of the `<menu>` element. I think the extension should do the same thing since it is semantically correct.

<img width="311" height="94" alt="Image" src="https://github.com/user-attachments/assets/b0c88896-b56b-4688-95aa-5655aa8f41ba" />

**Comment by straker** [comment]
A toolbar is a collection of controls. It doesn't have any required children and so any control can go inside of it. If you wan't to use `li` to group a list of controls, you should either use a `ul` as the parent of the list or have each `li` have `role=none/presentation`.

**Comment by micaavigliano** [comment]
got it, axe is the only tool that fails on this one, even voiceover is working correctly with the combination:

```
<menu role="toolbar">
  <li>
    <button>interaction</button>
  </li>
</menu>
```
https://github.com/user-attachments/assets/adf063c8-89e2-4e9c-ad21-fda639dea8d5

I do not know why I have to put a `role="presentation"` on the `<li>`, but I got it.

**Comment by WilcoFiers** [comment]
This is weirder than it looks on the surface @straker. Testing this out, @micaavigliano is right that putting role=toolbar on a menu turns the li's inside it into generics. I tested in Firefox and Chromium, and at least there that seems to be the case.

The trouble is, that this doesn't really make any sense. [ARIA in HTML](https://www.w3.org/TR/2026/REC-html-aria-20260407/) has the following to say about it:

> If the li is a child of a list element ([ul](https://www.w3.org/TR/2026/REC-html-aria-20260407/#el-ul), [ol](https://www.w3.org/TR/2026/REC-html-aria-20260407/#el-ol), [menu](https://www.w3.org/TR/2026/REC-html-aria-20260407/#el-menu)) role=[listitem](https://www.w3.org/TR/2026/REC-html-aria-20260407/#index-aria-listitem).
> 
> Otherwise, if the li is not a child of a list element it is exposed as a role=[generic](https://www.w3.org/TR/2026/REC-html-aria-20260407/#index-aria-generic).

This suggests to me that an `li` inside a `menu` is a `listitem` regardless of the role of the `menu` element. You could maybe argue it's implied that adding a role to `menu` would change the `li` to generic. I can sort of see that, and that does seem to be what Chromium and Firefox are doing. But then they aren't following that second part at all. `li` elements used outside a list element are still mapped to the listitem role. 

## Recommendations

I think we need to do more thorough testing to try to understand this behavior of changing the role on list elements.

1. Check combinations of menu, ul, ol, dl, div without role, and with role toolbar and generic, and with a child li, dd, dt
2. Test in Chrome, Edge, Safari, Firefox (Accessibility tree directly), record what role the child node had
2. Also check with screen readers directly (VO + Safari, NVDA + FF, JAWS + Chrome), record if the child was announced in some way

**Comment by pkra** [comment]
@WilcoFiers I think htm-aria needs an update here. 

The text you quoted is not normative but summarizes html-aam which says (https://w3c.github.io/html-aam/#el-li):

> If li element is not a child of [ol](https://html.spec.whatwg.org/multipage/grouping-content.html#the-ol-element) , [menu](https://html.spec.whatwg.org/multipage/grouping-content.html#menus) or [ul](https://html.spec.whatwg.org/multipage/grouping-content.html#the-ul-element), or if the containing list element is no longer exposed with a list role, then expose the li element with a generic role.

This matches what follows in html-aria:

> Otherwise, [any role](https://w3c.github.io/html-aria/#dfn-any-role) if the parent list element does not expose an implicit or explicit list role.

i.e. the typical language for role=generic.

I've filed https://github.com/w3c/html-aria/issues/577

**Comment by micaavigliano** [comment]
Hey, thanks for reopening the issue. I understand that the element `<menu>` is not very used and that is why there is a lot of doubts and room for learning and improving. This past week, I have been testing this element because to be honest I did not know it before. My five cents on this would be the following:
Semantically the `<menu>` alone, with no roles, has the same semantically value as a `<ul>` with the difference that it is not being used to group items but to group commands. By default, it shares the children with `<ul>` and, since it groups commands, it is expected that every command has a Tab stop. The thing changes when we add the `role="toolbar"` because now we are build a [Toolbar Pattern](https://www.w3.org/WAI/ARIA/apg/patterns/toolbar/) according to the W3 and it is expected to be navigated as a toolbar. 

This is the Accessibility Tree's node of the toolbar component.
<img width="292" height="95" alt="Image" src="https://github.com/user-attachments/assets/90e1fb4d-d793-47f4-9258-ebb5c9f8ec31" />

This is the HTML code:
<img width="407" height="447" alt="Image" src="https://github.com/user-attachments/assets/37707b3d-e6da-4f3d-b7bc-b87e468931c6" />

In this case, we will see that the the `<li>` role is omitted because it is not necessary for the `role="toolbar"` but expected on `<menu>`. Can we just add a `role="presentation"`? yes, of course, but it will be just patching to prevent an error that it is already handled by the [AAM 1.2 algorithm](https://www.w3.org/TR/html-aam/#el-li) when it says:

> If li element is not a child of [ol](https://html.spec.whatwg.org/multipage/grouping-content.html#the-ol-element) , [menu](https://html.spec.whatwg.org/multipage/grouping-content.html#menus) or [ul](https://html.spec.whatwg.org/multipage/grouping-content.html#the-ul-element), or if the containing list element is no longer exposed with a list role, then expose the li element with a generic role.

_"f the containing list element is no longer exposed with a list role, then expose the li element with a generic role"_ , in this case is no longer a menu but a toolbar and that is why is being omitted by the accessibility tree.

Again, thanks for reconsidering!!

**Comment by WilcoFiers** [comment]
@pkra Thanks for raising this to ARIA in HTML. That would be useful to clarify. But the more important problem I think is actually in AAM. An `li` used anywhere other then as a child of `ul`, `ol`, `menu` is still mapped to `listitem`. That's why we missed this case in the first place `div > li` turns into `generic > listitem` in the accessibility tree, even though AAM and ARIA in HTML are suggesting this should be `generic > generic`.

**Comment by pkra** [comment]
@WilcoFiers right. I noticed that when reviewing Scott's PR https://github.com/w3c/html-aria/pull/578 just now. 

So I also filed https://github.com/w3c/html-aam/issues/604. (But I'm not sure how much HTML-AAM is willing or able to say about invalid HTML markup like `div > li`.)

**Comment by micaavigliano** [comment]
I would not consider this as an html issue per se because the `role=“toolbar”` is not expecting an item as a child but a command. If you see on the video I shared about the behavior of the element is correct, it announced how many items the group has and it traps the navigation as the toolbar pattern in w3 says. 
The main issue is that axe extension is not recognizing `<menu role=“toolbar”> > li` as a valid pattern

**Comment by WilcoFiers** [comment]
> So I also filed https://github.com/w3c/html-aam/issues/604. (But I'm not sure how much HTML-AAM is willing or able to say about invalid HTML markup like div > li.)

Interesting! Maybe that's why browsers are doing it this way. They coded up the `generic` logic, but only applied it to valid HTML scenarios. That explains what we're seeing.

**Comment by pkra** [comment]
@WilcoFiers I realized just now that https://github.com/w3c/aria/pull/2212 is still waiting for implementations. Sigh.

**Comment by micaavigliano** [comment]
Yes, absolutely! thanks for the quick response and updating the w3c! It is amazing the work you are doing and having the space to open this discussions is great.

(btw, in my current work we are using axe for almost everything)
